### PR TITLE
Onboarding Customisation Experiment

### PR DESCRIPTION
**Pull request by nalcalag**
Task/Issue URL: https://app.asana.com/0/72649045549333/1203826977705495/f

### Description
Implement Onboarding Customisation experiment where experimental group will see an extra onboarding dialog with DDG feature options.

### Steps to test this PR

_Change all variants to 0 except for the `mj` variant_

_New multiselect dialog_
- Uninstall previous version of the app or clear storage
- Install from this branch
- Welcome animation will appear
- [ ] Check new dialog with DDG feature options appear

_Skip selecting options_
- Follow same instructions as "New multiselect dialog" test
- Tap on `Skip This Step`
- [ ] Check existing welcome cta dialog appears
- [ ] Check pixel `m_onboarding_option_skip` is fired

_Continue without options selected_
- Follow same instructions as "New multiselect dialog" test
- Tap on any option to select it
- Remove the selection tapping again
- Tap on `Continue` button
- [ ] Check existing welcome cta dialog appears
- [ ] Check pixel `m_onboarding_option_skip` is fired

_Continue with options selected_
- Follow same instructions as "New multiselect dialog" test
- Tap on one or more options to select them
- Tap on `Continue` button
- [ ] Check existing welcome cta dialog appears
- [ ] Check pixel `m_onboarding_option_selected` is fired
- [ ] Check pixels for each option selected is also fired. You can find the list of pixels for each option here -> https://app.asana.com/0/0/1204137437551040/f

### UI changes

https://user-images.githubusercontent.com/20798495/228512221-df3caf98-2bda-4f0e-9976-adfc51776f9e.mp4

### CODE REVIEW

**Comment by nalcalag** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3003** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3003" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3003?utm_source=stack-comment).

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePage.kt [hunk 1]
```diff
@@ -189,14 +250,20 @@ class WelcomePage : OnboardingPageFragment(R.layout.content_onboarding_welcome)
...
-                        welcomeAnimationFinished = true
-                        binding.daxDialogCta.dialogTextCta.startTypingAnimation(ctaText)
-                        setPrimaryCtaListenerAfterWelcomeAlphaAnimation()
-                    }
+                event(WelcomePageView.Event.ShowFirstDaxOnboardingDialog)
+            }
+    }
+
+    private fun showDaxDialogCta() {
+        binding.daxDialogCta.root.show()
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePage.kt:258]: same question about `root` here. Is this something related to having a `include` in the layout?

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt [hunk 2]
```diff
@@ -33,16 +42,52 @@ class WelcomePageViewModel(
...
+
+    private fun onSkipOptionsClicked(): Flow<WelcomePageView.State> = flow {
+        pixel.fire(AppPixelName.ONBOARDING_OPTION_SKIP)
+        emit(WelcomePageView.State.ShowControlDaxCta)
+    }
+
+    private fun onContinueWithOptionsClicked(options: Map<DDGFeatureOnboardingOption, Boolean>): Flow<WelcomePageView.State> = flow {
+        val optionsSelected = options.filter { it.value }.map { it.key }.toList()
+        if (optionsSelected.isEmpty()) {
+            pixel.fire(AppPixelName.ONBOARDING_OPTION_SKIP)
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt:74]: I thought this was not possible, but testing I see it's possible to select and unselect again any option, leaving the primary button enabled without anything selected. I assume we want that, but sharing I would have expected the primary to be disabled again.

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt [hunk 3]
```diff
@@ -33,16 +42,52 @@ class WelcomePageViewModel(
...
+            false -> emit(WelcomePageView.State.ShowControlDaxCta)
+        }
+    }
+
+    private fun onSkipOptionsClicked(): Flow<WelcomePageView.State> = flow {
+        pixel.fire(AppPixelName.ONBOARDING_OPTION_SKIP)
+        emit(WelcomePageView.State.ShowControlDaxCta)
+    }
+
+    private fun onContinueWithOptionsClicked(options: Map<DDGFeatureOnboardingOption, Boolean>): Flow<WelcomePageView.State> = flow {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt:71]: NIT: if we ever decide to release the experiment (I don't think is the plan), I would move all the pixel logic into its own class.

File: app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt [hunk 4]
```diff
@@ -40,6 +40,15 @@ enum class AppPixelName(override val pixelName: String) : Pixel.PixelName {
     ONBOARDING_DAX_CTA_OK_BUTTON("m_odc_ok"),
     ONBOARDING_PRIVACY_SHIELD_BUTTON("m_onboarding_privacy_shield_button"),
 
+    ONBOARDING_OPTION_PRIVATE_SEARCH_SELECTED("m_onboarding_option_private_search_selected"),
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt:43]: did we open a privacy triage for these ones?

File: app/src/main/res/layout/include_dax_dialog_cta.xml
Comment (cmonfortep) [inline app/src/main/res/layout/include_dax_dialog_cta.xml:1]: probably not part of the experiment and we can discard changes here?

File: app/src/main/res/layout/include_dax_multiselect_dialog_cta.xml [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/daxMultiselectCtaContainer"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:alpha="0"
+    android:clipChildren="false"
+    android:orientation="vertical"
+    android:visibility="gone">
```
Comment (cmonfortep) [inline app/src/main/res/layout/include_dax_multiselect_dialog_cta.xml:25]: NIT: I would avoid setting visibility to something we can include somewhere else.

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt [hunk 6]
```diff
@@ -23,17 +23,33 @@ import com.duckduckgo.app.global.DispatcherProvider
...
     private val userStageStore: UserStageStore,
     private val pageLayoutManager: OnboardingPageManager,
     private val dispatchers: DispatcherProvider,
+    private val variantManager: VariantManager,
 ) : ViewModel() {
 
+    private val command = Channel<Command>(1, DROP_OLDEST)
+    internal fun commands(): Flow<Command> = command.receiveAsFlow()
+
+    fun determineScreenOrientation() {
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt:47]: include one test for this?

**Review by cmonfortep — APPROVED** [review]
LGTM

works as expected, leaving the UI discussion to the UI code owners 😁.

(Agree with David that new components should go through ADS standard process, however not clear if this is an established component atm. I assume we have some lint checks and that's why we put some colors inside common-ui, if that's the case, maybe we can do something similar to what we do with strings...have a separated file, or declare them under a comment that indicates they are not established. Just some thoughts)

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt [hunk 7]
```diff
@@ -52,4 +68,8 @@ class OnboardingViewModel @Inject constructor(
             userStageStore.stageCompleted(AppStage.NEW)
         }
     }
+
+    internal sealed class Command {
+        object SetScreenOrientation : Command()
```
[the remark this answers is not in the document]
Comment (nalcalag) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt:73]: You are right, I changed it to `ForceToPortraitForMobileDevices` so it's more clear 👍

File: app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt (hunk 4 above)
Comment (nalcalag) [reply app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt:43]: yes :)

File: app/src/main/res/layout/include_dax_dialog_cta.xml
Comment (nalcalag) [reply app/src/main/res/layout/include_dax_dialog_cta.xml:1]: 💯

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt (hunk 6 above)
Comment (nalcalag) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingViewModel.kt:47]: I will 👌

File: app/src/main/res/layout/include_dax_multiselect_dialog_cta.xml (hunk 5 above)
Comment (nalcalag) [reply app/src/main/res/layout/include_dax_multiselect_dialog_cta.xml:25]: You are right, I will move this logic to the class 💯

**Comment by nalcalag** [comment]
@malmstein following @cmonfortep suggestion I added a new resource file called `temp_colors.xml` where I put all the colors needed for the customView because I need to add them to the theme file so they can change depending on the theme. I also migrated all styles and attributes to new resources files in the `app` module so they are not in `common-ui` anymore 🎉

@cmonfortep thanks a lot for all your feedback, I changed everything you suggested and added tests for portrait checks 👍

**Review by malmstein — APPROVED** [review]
Thanks for doing the changes @nalcalag. Good work!
